drivers/gpio/gpio_utils.h

```diff
@@ -18,19 +18,27 @@
  * @param callbacks A pointer to the original list of callbacks (can be NULL)
  * @param callback A pointer of the callback to insert or remove from the list
  * @param set A boolean indicating insertion or removal of the callback
+ *
+ * @return 0 on success, negative errno otherwise.
  */
-static inline void _gpio_manage_callback(sys_slist_t *callbacks,
-					 struct gpio_callback *callback,
-					 bool set)
+static inline int _gpio_manage_callback(sys_slist_t *callbacks,
+					struct gpio_callback *callback,
+					bool set)
 {
 	__ASSERT(callback, "No callback!");
 	__ASSERT(callback->handler, "No callback handler!");
 
+	if (!sys_slist_is_empty(callbacks)) {
+		if (!sys_slist_find_and_remove(callbacks, &callback->node)) {
+			return -EINVAL;
```

Comment:
shouldn't this return `-EINVAL` only if `set` is false?

Reply:
no, we also need to know if it's already present in the list if we want to add it. Since there is not sys_slist_find() I just used sys_slist_find_and_remove(), and below if set: it's added again.

Reply:
Sorry, still confused.  I read the code as "if there's already callbacks, and when we try to remove the new one it isn't there, return -EALREADY".  So how can you add a new callback to a list that isn't empty and doesn't already have the callback in it?

Reply:
ah ... ouch :) . Was so much narrow minded on the bug's use case I forgot that you cannot check if an actual node is already in a list. Actually, that's not fixable then. (though checking if the node is there before removing it is still fine). will fix